Load Yang model default value to DefaultValueProvider

liuh-80 opened the pull request:
#### Why I did it
Support sonic-swss-common read Yang model defaut value.

#### How I did it
Load Yang model default value to DefaultValueProvider.

#### How to verify it
Add new UT to cover DefaultValueProvider code.
Pass all UT.

#### Which release branch to backport (provide reason below if selected)

<!--
- Note we only backport fixes to a release branch, *not* features!
- Please also provide a reason for the backporting below.
- e.g.
- [x] 202006
-->

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

#### Description for the changelog
Load Yang model default value to DefaultValueProvider.

#### Link to config_db schema for YANG module changes
<!--
Provide a link to config_db schema for the table for which YANG model
is defined
Link should point to correct section on https://github.com/Azure/SONiC/wiki/Configuration.
-->

#### A picture of a cute animal (not mandatory but encouraged)

liuh-80 commented on .azure-pipelines/build-sairedis-template.yml:
```diff
@@ -20,6 +20,10 @@ parameters:
 - name: sonic_slave
   type: string
 
```
All azure pipeline change is from another PR: https://github.com/sonic-net/sonic-swss-common/pull/681
That PR need merge first.

liuh-80 commented:
/azp run

liuh-80 commented:
/azp run

qiluo-msft commented:
> #include <fstream>

You removed the whole content and changed file attribute. Is it intended?

---
Refers to: common/dbconnector.cpp:7 in 00935a3. [](commit_id = 00935a38360083bcc5ea7bbbb073e947985e60e9, deletion_comment = True)

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+    _Pragma ("GCC diagnostic push") \
+    _Pragma ("GCC diagnostic ignored \"-Wcast-align\"")
+#define WARNINGS_RESET \
+   _Pragma ("GCC diagnostic pop")
```
>    [](http://example.com/codeflow?start=0&length=3)

Indent 4 spaces?

liuh-80 replied:
Fixed.

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+    // C++ need this empty ctor
+}
+
+shared_ptr<string> TableInfoBase::GetDefaultValue(const string &key, const string &field)
```
>shared_ptr<string> [](http://example.com/codeflow?start=0&length=18)

Just return a `string`.

liuh-80 replied:
Can't return string because following 2 scenario are different:
1. When Yang model set default value to a empty string, we need return a empty string.
2. When Yang model does not set default value, we need return a nullptr.

qiluo-msft commented on common/defaultvalueprovider.h:
```diff
@@ -0,0 +1,118 @@
+    std::shared_ptr<std::string> GetDefaultValue(const std::string &row, const std::string &field);
+
+protected:
+    virtual bool FindFieldMappingByKey(const std::string &row, FieldDefaultValueMapping ** foundedMappingPtr) = 0;
```
>row [](http://example.com/codeflow?start=58&length=3)

do you mean `key`?

liuh-80 replied:
Fixed, also fix other similer cases.

qiluo-msft commented on common/defaultvalueprovider.h:
```diff
@@ -0,0 +1,118 @@
+public:
+    static DefaultValueProvider& instance();
+
+    void appendDefaultValues(const std::string &table, const std::string &key, std::vector<std::pair<std::string, std::string>> &values);
```
>values [](http://example.com/codeflow?start=129&length=6)

Do you mean `fvs`? ie. field and value pairs

liuh-80 replied:
Fixed.

qiluo-msft commented on common/defaultvalueprovider.h:
```diff
@@ -0,0 +1,118 @@
+    DefaultValueProvider();
+    ~DefaultValueProvider();
+
+    //  libyang context
```
>   [](http://example.com/codeflow?start=6&length=2)

One blank?

liuh-80 replied:
Fixed.

qiluo-msft commented on common/defaultvalueprovider.h:
```diff
@@ -0,0 +1,118 @@
+
+    void Initialize(char* modulePath = DEFAULT_YANG_MODULE_PATH);
+    
+    void LoadModule(const std::string name, const std::string path, struct ly_ctx *context);
```
>const std::string [](http://example.com/codeflow?start=20&length=17)

Normally use `string` or `const string&`.

liuh-80 replied:
Fixed, change to 'const string&'

qiluo-msft commented on tests/defaultvalueprovider_ut.cpp:
```diff
@@ -0,0 +1,98 @@
+TEST(DECORATOR, GetDefaultValues)
+{
+    DefaultValueProvider instance;
+    instance.Initialize("./tests/yang");
```
>./tests/yang [](http://example.com/codeflow?start=25&length=12)

The relative path is working only when the cwd is this "tests" folder.

liuh-80 replied:
There is no better solution, other C++ UT also using relative path for config file.

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+    }
+}
+
+bool TableInfoDict::FindFieldMappingByKey(const string &key, FieldDefaultValueMapping ** foundedMappingPtr)
```
>bool [](http://example.com/codeflow?start=0&length=4)

If return a bool, recommend the function name `FoundFieldMappingByKey`.

`FindXxx` normally return an iterator.

liuh-80 replied:
Fixed, change to FoundFieldMappingByKey

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+// existedValues and targetValues can be same container.
+void TableInfoBase::AppendDefaultValues(const string &row, map<string, string>& existedValues, map<string, string>& targetValues)
+{
+    assert(!row.empty());
```
>assert [](http://example.com/codeflow?start=4&length=6)

This is a public function, you should not use `assert` to validate input because it is noop in release build.

liuh-80 replied:
Fixed, assert removed.

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+    }
+}
+
+bool TableInfoDict::FindFieldMappingByKey(const string &key, FieldDefaultValueMapping ** foundedMappingPtr)
```
>foundedMappingPtr [](http://example.com/codeflow?start=89&length=17)

typo: foundMappingPtr

liuh-80 replied:
Fixed.

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+
+    SWSS_LOG_DEBUG("TableInfoDict::FindFieldMappingByKey %s\n", key.c_str());
+    auto keyResult = m_defaultValueMapping.find(key);
+    *foundedMappingPtr = keyResult->second.get();
```
>keyResult [](http://example.com/codeflow?start=25&length=9)

If not found, will it throw?

liuh-80 replied:
When not found, here will not throw exception.
When not found, the keyResult->second.get() will return a nullptr and the return value of this method will be False. the caller code will check return value to handle this case.

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+    int fieldCount = (int)count(key.begin(), key.end(), '|') + 1;
+    auto keyInfo = m_defaultValueMapping.find(fieldCount);
+    
+    // when not found, key_info still a valied iterator
```
>valied [](http://example.com/codeflow?start=40&length=6)

typo

liuh-80 replied:
Fixed.

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+DefaultValueProvider& DefaultValueProvider::instance()
+{
+    static DefaultValueProvider instance;
+#ifndef UNITTEST
```
>UNITTEST [](http://example.com/codeflow?start=8&length=8)

Why you need this? Possible to remove unit test related conditional code?

liuh-80 replied:
Will check if gmock can fix this issue and update later.

liuh-80 replied:
Fixed.

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+            SWSS_LOG_DEBUG("list field: %s\n",field->name);
+            GetDefaultValueInfoForLeaflist(field, fieldMapping);
+        }
+#ifdef DEBUG
```
>DEBUG [](http://example.com/codeflow?start=7&length=5)

Why need this check?

liuh-80 replied:
Fixed.

qiluo-msft commented on common/defaultvalueprovider.h:
```diff
@@ -0,0 +1,118 @@
+struct ly_ctx;
+
+// Key information
+typedef std::pair<std::string, int> KeyInfo;
```
>KeyInfo [](http://example.com/codeflow?start=36&length=7)

I checked HLD and there is no document. Suggest add enough code comments to explain the struct. It is difficult to understand:
1. what is key's infor?
2. what is the first string?
3. what is the second int?

liuh-80 replied:
Fixed, add more detail to comments.

qiluo-msft commented on common/defaultvalueprovider.h:
```diff
@@ -0,0 +1,118 @@
+typedef std::pair<std::string, int> KeyInfo;
+
+// Field name to default value mapping
+typedef std::map<std::string, std::string> FieldDefaultValueMapping;
```
>string [](http://example.com/codeflow?start=35&length=6)

Is it true that the default value is not a string?
For example: [https://github.com/sonic-net/sonic-buildimage/blob/19155df148fbb49eecd2ce4a3b8f022a66beab76/src/sonic-yang-models/yang-models/sonic-buffer-pg.yang#L54](https://github.com/sonic-net/sonic-buildimage/blob/19155df148fbb49eecd2ce4a3b8f022a66beab76/src/sonic-yang-models/yang-models/sonic-buffer-pg.yang#L54)

liuh-80 replied:
Yes, it's true that default value may not a string.

However, libyang always get default value as string. so in internal we store the default value as string.

Also currently swss-common will return field-value pair as a string to string dictionary.

qiluo-msft commented on common/defaultvalueprovider.h:
```diff
@@ -0,0 +1,118 @@
+    ~DefaultValueProvider();
+
+    //  libyang context
+    struct ly_ctx *m_context = nullptr;
```
>m_context [](http://example.com/codeflow?start=19&length=9)

You are assuming there is only one `ly_ctx`. Is it always true? Can we relax?

liuh-80 replied:
It's true, there will only one ly_ctx, it's created by our code and this is a singleton.
Libyang only need 1 context, there is no scenario that we need multiple context.

qiluo-msft replied:
Today the answer is yes. I am think relax the limitation may be future proof.

In future, ApplDB may have separate yang models, structure events may also has separate yang models. And these yang models are independent with each other. So I guess there will be use case that clients may use multiple `ly_ctx.`.

The lazy initialization is still useful since the client of swsscommon may not use yang features at all.

liuh-80 replied:
Fixed, now defaultvalue provider is not a singleton, so we can use multiple default value provider to load default value for different DB.

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+    }
+}
+
+FieldDefaultValueMappingPtr DefaultValueProvider::GetDefaultValueInfo(struct lys_node* tableChildNode)
```
>GetDefaultValueInfo [](http://example.com/codeflow?start=50&length=19)

This function has nothing todo with the concept of "default value". Seems a general help function related to yang data structure. Suggest move out of this class.

liuh-80 replied:
Fixed, create a new DefaultValueHelper class and move these method to DefaultValueHelper and change to static method.

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+        if (field->nodetype == LYS_LEAF)
+        {
+            SWSS_LOG_DEBUG("leaf field: %s\n",field->name);
+            GetDefaultValueInfoForLeaf(field, fieldMapping);
```
>field [](http://example.com/codeflow?start=39&length=5)

You just checked the `nodetype` field, so better cast `field` to proper class in this function, and no need to cast in the lower level function such as `GetDefaultValueInfoForLeaf`.

liuh-80 replied:
Fixed

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,544 @@
+#include "table.h"
+#include "json.h"
+
+#if defined(__arm__) || defined(__aarch64__)
```
>__aarch64__ [](http://example.com/codeflow?start=32&length=11)

SONiC will be built on arm platforms. Is it safe to ignore the warning (as error)?

liuh-80 replied:
New UT cover the warning code and passed, so I think it's safe.
Also I check this issue online and found there some discussion suggest ignore this warning is safe.

liuh-80 commented:
> > #include
> 
> You removed the whole content and changed file attribute. Is it intended?
> 
> Refers to: common/dbconnector.cpp:7 in [00935a3](https://github.com/sonic-net/sonic-swss-common/commit/00935a38360083bcc5ea7bbbb073e947985e60e9). [](commit_id = [00935a3](https://github.com/sonic-net/sonic-swss-common/commit/00935a38360083bcc5ea7bbbb073e947985e60e9), deletion_comment = True)

Fixed, this PR not change any content in dbconnector.cpp, I only fix the incorrect file mode issue. seems a github bug.

qiluo-msft commented on common/defaultvalueprovider.cpp:
```diff
@@ -0,0 +1,493 @@
+#if defined(__arm__) || defined(__aarch64__)
+#define WARNINGS_NO_CAST_ALIGN \
+     _Pragma ("GCC diagnostic push") \
+     _Pragma ("GCC diagnostic ignored \"-Wcast-align\"")
```
>      [](http://example.com/codeflow?start=0&length=5)

Indent 4 spaces? 5 is weird.

liuh-80 replied:
Fixed

qiluo-msft commented:
> #include <fstream>

I see non-critical comment change and still file attribute change. If it is intended change, suggest move to a standalone PR. This PR is complex and should be focused.

---
In reply to: [1242820539](https://github.com/sonic-net/sonic-swss-common/pull/682#issuecomment-1242820539) [](http://example.com/codeflow?ancestors=1242820539,1242820539)

---
Refers to: common/dbconnector.cpp:7 in 00935a3. [](commit_id = 00935a38360083bcc5ea7bbbb073e947985e60e9, deletion_comment = True)

liuh-80 commented:
> > #include
> 
> I see non-critical comment change and still file attribute change. If it is intended change, suggest move to a standalone PR. This PR is complex and should be focused.
> 
> In reply to: [1242820539](https://github.com/sonic-net/sonic-swss-common/pull/682#issuecomment-1242820539)
> 
> Refers to: common/dbconnector.cpp:7 in [00935a3](https://github.com/sonic-net/sonic-swss-common/commit/00935a38360083bcc5ea7bbbb073e947985e60e9). [](commit_id = [00935a3](https://github.com/sonic-net/sonic-swss-common/commit/00935a38360083bcc5ea7bbbb073e947985e60e9), deletion_comment = True)

Fixed, reverted change in this file.

liuh-80 commented:
/azp run

liuh-80 commented:
Currently, there still 10 repos need update their azure pipeline to install libyang. 10 PRs already created.
This PR need wait for those PRs finish first.
